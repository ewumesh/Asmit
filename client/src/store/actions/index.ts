import * as authorizeActions from './authorizeActions'
import * as globalActions from './globalActions'
import * as imageGalleryActions from './imageGalleryActions'
import * as userActions from './userActions'
import * as localeActions from './localeActions'
import * as serverActions from './serverActions'
import * as productActions from './productActions'
import * as shippingsActions from './shippingsActions'
import * as addToCartActions from './addToCartActions'

export {
  authorizeActions,
  globalActions,
  imageGalleryActions,
  userActions,
  localeActions,
  serverActions,
  productActions,
  shippingsActions,
  addToCartActions
}
