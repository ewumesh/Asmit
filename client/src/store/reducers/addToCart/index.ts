import { addToCartReducer } from './addToCartReducer'
export {
    addToCartReducer
}
