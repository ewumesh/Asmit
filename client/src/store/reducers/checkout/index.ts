import { checkoutReducer } from './checkoutReducer'
export {
    checkoutReducer
}
