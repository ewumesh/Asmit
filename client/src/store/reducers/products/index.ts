import { productReducer } from './productReducer'
import { productSelector } from './productSelector'

export {
    productReducer,
    productSelector
}
