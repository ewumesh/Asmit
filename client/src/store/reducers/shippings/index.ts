import { shippingsReducer } from './shippingsReducer'
import { shippingsSelector } from './shippingsSelector'

export {
    shippingsReducer,
    shippingsSelector
}
