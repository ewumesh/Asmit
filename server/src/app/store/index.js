module.exports = {
  AddToCart: require('./AddToCart'),
  GetCart: require('./GetCart'),
  PlaceOrder: require('./PlaceOrder'),
  UpdateOrder: require('./UpdateOrder'),
  UpdateCartItem: require('./UpdateCartItem'),
};
