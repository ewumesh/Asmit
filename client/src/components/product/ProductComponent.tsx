// - Import react components
import { Map } from 'immutable'
import { NavLink } from 'react-router-dom'
import { connect } from 'react-redux'
import { getTranslate } from 'react-localize-redux'
import { push } from 'connected-react-router'
import { withStyles } from '@material-ui/core/styles'
import Grid from '@material-ui/core/Grid/Grid'
import Paper from '@material-ui/core/Paper'
import React, { Component } from 'react'
import Typography from '@material-ui/core/Typography'

import { Cart } from 'core/domain/cart'
import * as addToCartActions from 'store/actions/addToCartActions'

import { IProductComponentProps } from './IProductComponentProps'
import { IProductComponentState } from './IProductComponentState'
import FormButton from '../formButton'
import uniqid from 'uniqid'

const styles = (theme: any) => ({
  image: {
    verticalAlign: 'top',
    maxWidth: '100%',
    minWidth: '100%',
    width: '100%'
  },
  paper: {
    padding: theme.spacing(2),
    height: '100%',
    color: theme.palette.text.secondary,
    textAlign: 'center'
  },
})

/**
 * Create component class
 */
export class ProductComponent extends Component<IProductComponentProps,IProductComponentState> {

  styles = {
  }

  /**
   * Component constructor
   * @param  {object} props is an object properties of component
   */
  constructor (props: IProductComponentProps) {
    super(props)

    // Defaul state
    this.state = {
    }

    // Binding functions to `this`
    this.handleBuyNow = this.handleBuyNow.bind(this)

  }

  /**
   * Will be called when buynow button is pressed
   *
   * @memberof Product
   */
  handleBuyNow = (product: any) => {
    const { authed, goTo } = this.props
     if (!authed) {
       goTo!('/login')
       return
     }

     const{cart} = this.props
     const cartProduct: any[] = []
     let cartId
     var cartSubTotal = 0
     if (cart && Object.keys(cart).length > 0) {
       cartId = Object.keys(cart)[0]
       console.log('cart id coming while updatin the cart',cartId)
     } else {
       cartId = uniqid()
     }

    this.props.addToCart!(cartId,{
      productThumbnail: product.get('thumbnail'),
      productName: product.get('name'),
      productPrice: product.get('price'),
      productId: product.get('productId'),
      productColor: 'blue',
      productQuantity: 1,
      productSize: 'L'
    })
  }

  /**
   * Reneder component DOM
   * @return {react element} return the DOM which rendered by component
   */
  render () {

    let { product, goTo, cart } = this.props
    const {classes, translate} = this.props
    return (
        <Grid key={product.get('productId')} item xs={3}>
          <Paper className={classes.paper}>
          <NavLink
                    to={{
                        pathname: `/productDetail/` + product.get('productId'),
                    }}
                    onClick={evt => {
                        evt.preventDefault()
                        goTo!(`/productDetail/` + product.get('productId'))
                    }}
                >
            <img
              className={classes.img}
              alt='complex'
              src={`/images/${product.get('thumbnail')}`}
            />
            <Typography variant='subtitle1' style={{ textAlign: 'center' }}>
              {product.get('name')}
            </Typography>
            <Typography color='error' style={{ textAlign: 'center' }}>
              {translate!('product.currency')}{product.get('price')}
            </Typography>
            </NavLink>
            <FormButton
              size='small'
              align='center'
              color='secondary'
              onClick={() => this.handleBuyNow(product)}
            >
              {translate!('product.buynow')}
            </FormButton>
          </Paper>
        </Grid>
    )
  }
}

/**
 * Map dispatch to props
 * @param  {func} dispatch is the function to dispatch action to reducers
 * @param  {object} ownProps is the props belong to component
 * @return {object}          props of component
 */
const mapDispatchToProps = (dispatch: any, ownProps: IProductComponentProps) => {
  return {
    goTo: (url: string) => dispatch(push(url)),
    addToCart: (cartId: string, cart: Cart) => {
      dispatch(addToCartActions.dbAddProductToCart(cartId,cart))
    }
  }
}

/**
 * Map state to props
 * @param  {object} state is the obeject from redux store
 * @param  {object} ownProps is the props belong to component
 * @return {object}          props of component
 */
const mapStateToProps = (state: Map<string, any>, ownProps: IProductComponentProps) => {
  return {
    translate: getTranslate(state.get('locale')),
    avatarURL: state.getIn(['imageGallery', 'imageURLList']),
    authed: state.getIn(['authorize', 'authed'], false),
    imageRequests: state.getIn(['imageGallery', 'imageRequests']),
    cart: state.getIn(['addToCart', 'cartProducts']).toJS()
  }
}

// - Connect component to redux store
export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles as any)(ProductComponent as any)as any)
