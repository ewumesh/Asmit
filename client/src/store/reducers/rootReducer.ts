import { localeReducer as locale } from 'react-localize-redux'
import {
  combineReducers
} from 'redux-immutable'

// - Import reducers
import { authorizeReducer } from './authorize'
import { globalReducer } from './global'
import { imageGalleryReducer } from './imageGallery'
import { userReducer } from './users'
import { serverReducer } from './server'
import { productReducer } from './products'
import { shippingsReducer } from './shippings'
import { checkoutReducer } from './checkout'
import { addToCartReducer } from './addToCart'
import { connectRouter } from 'connected-react-router/immutable'

// - Reducers
export const rootReducer = (history: any) => combineReducers({
    locale,
    imageGallery: imageGalleryReducer,
    server: serverReducer,
    authorize: authorizeReducer,
    router: connectRouter(history),
    user: userReducer,
    products: productReducer,
    shippings: shippingsReducer,
    checkout: checkoutReducer,
    addToCart: addToCartReducer,
    global: globalReducer
  } as any)
