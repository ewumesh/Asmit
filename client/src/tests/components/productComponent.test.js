import React from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import createHistory from 'history/createBrowserHistory'
import { rootReducer } from 'src/store/reducers'

import ProductComponent from 'src/components/product/ProductComponent';
import { Map, fromJS } from 'immutable';

// create any initial state needed
const initialState = Map({

});

const mockStore = configureMockStore();
const store = mockStore(initialState);
import '../../setupTests';

it('Renders correctly', () => {
  const wrapper = shallow(
    <Provider store={store}>
      <ProductComponent
        product={Map({
          productId: 1,
          name:
            "Arc d'Triomphe,description:This beautiful and iconic T-shirt will no doubt lead you to your own triumph",
          price: 14.99,
          discountedPrice: 0,
          image: 'arc-d-triomphe.gif',
          image2: 'arc-d-triomphe-2.gif',
          thumbnail: 'arc-d-triomphe-thumbnail.gif',
          display: 0
        })}
      />
    </Provider>
  );
  expect(wrapper).toMatchSnapshot();
});
