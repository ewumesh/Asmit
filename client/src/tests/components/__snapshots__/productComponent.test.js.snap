// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeState": Immutable.Map {
    },
    }
  }
>
  <Connect(WithStyles(ProductComponent))
    product={
      Immutable.Map {
        productId: 1,
        name: "Arc d'Triomphe,description:This beautiful and iconic T-shirt will no doubt lead you to your own triumph",
        price: 14.99,
        discountedPrice: 0,
        image: "arc-d-triomphe.gif",
        image2: "arc-d-triomphe-2.gif",
        thumbnail: "arc-d-triomphe-thumbnail.gif",
        display: 0,
      }
    }
  />
</ContextProvider>
`;
